src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.cs

```diff
@@ -44,6 +46,69 @@ public string[] UserLanguages
             }
         }
 
+        private CookieCollection ParseCookies(Uri uri, string setCookieHeader)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Info(this, "uri:" + uri + " setCookieHeader:" + setCookieHeader);
+            CookieCollection cookies = new CookieCollection();
+            CookieParser parser = new CookieParser(setCookieHeader);
+            for (;;)
+            {
+                Cookie cookie = parser.GetServer();
+                if (NetEventSource.IsEnabled) NetEventSource.Info(this, "CookieParser returned cookie:" + cookie?.ToString());
+                if (cookie == null)
+                {
+                    // EOF, done.
+                    break;
+                }
+                if (cookie.Name.Length == 0)
+                {
+                    continue;
+                }
+
+                // This is a modified copy of CoookieCollection.InternalAdd, which is internal to System.Net.Primitives.
+                // It differs from the original in some subtle ways.
+                // Firstly, new Variant cookies will not overwrite older ones. We can't have this behaviour as Cookie.Variant
+                // is internal. This violates the Cookie2 spec.
+                // Secondly, if the Cookie version is not the MaxSupportedVersion, we don't set the CookieCollection.IsOtherVersionSeen
+                // to true. This causes different behaviour when calling CookieContainer.GetCookieHeader. This doesn't matter, as we are
+                // just using CookieCollection (not CookieContainer) here.
+                bool cookieExists = false;
+                foreach (Cookie addedCookie in cookies)
+                {
+                    if (CookieComparer.Compare(cookie, addedCookie) == 0)
```

Comment:
This is a copy of the code of CookieCollection.InternalAdd that I can't use because, well, it's internal. CookieCollection.InternalAdd uses this comparer for some reason. It'd be a backwards incompatible change to use cookie.Equals  as I think cookie.Equals would return false for things cookiecomparer would return true.